torch/_refs/__init__.py

```diff
@@ -2988,6 +2989,63 @@ def diagonal(
     return result
 
 
+@register_decomposition(torch.ops.aten.diag_embed)
+def diag_embed(
+    t: TensorLikeType,
+    offset: int = 0,
+    dim1: int = -2,
+    dim2: int = -1,
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.diag_embed
+    """
+    # as per the docs, exchanging dims is equivalent to changing the sign of
+    # offset
+    if dim1 > dim2:
+        dim1, dim2 = dim2, dim1
+        offset = -offset
+
+    # convert from negative dims
+    rank = t.ndim + 1
+    dim1 = utils.canonicalize_dim(rank=rank, idx=dim1)
+    dim2 = utils.canonicalize_dim(rank=rank, idx=dim2)
+
+    check(
+        dim1 != dim2, lambda: f"diagonal dimensions cannot be identical {dim1}, {dim2}"
+    )
+
+    # as per the docs, the size of last dim is placed at dim1 and dim2
+    last_dim = t.size(-1)
+
+    if offset != 0:
+        # add padding to match the new size
+        t_shape = list(t.shape)
+        t_shape[-1] = builtins.abs(offset)
+        z = torch.zeros(t_shape, dtype=t.dtype, device=t.device, requires_grad=False)
+        pair = (z, t) if offset > 0 else (t, z)
+        t = torch.cat(pair, dim=-1)
+        # make sure the diagonal always has the same size
+        last_dim += builtins.abs(offset)
+
+    # preserve original data, but place 1 at dim1 and move last dim to dim2
+    t = t.unsqueeze(dim1).movedim(-1, dim2)
+
+    # generate ranges shifting indices based on offset
+    a_range = torch.arange(last_dim, device=t.device, dtype=torch.int64)
+    b_range = torch.arange(
+        offset, last_dim + offset, device=t.device, dtype=torch.int64
+    )
+
+    # broadcast
+    cond = a_range == b_range.unsqueeze(-1)
+    cond_shape = [last_dim if i in (dim1, dim2) else 1 for i in range(len(t.shape))]
+    cond = cond.reshape(cond_shape)
+    zero = torch.zeros(1, dtype=t.dtype, device=t.device, requires_grad=False)
+    result = torch.where(cond, t, zero)
```

Comment:
if it's just mypy complaining, just silence it (look for other examples), no reason to create extra tensors when it should work already.

Reply:
i've realized this works, so changing to this (will change for eye as well):

```py
    result = torch.where(cond, t, False)
```

before that, i had this:

```py
    result = torch.where(cond, t, 0)
```

which resulted in these errors:
```
FAILED test/test_ops.py::TestCommonCPU::test_python_ref__refs_diag_embed_cpu_bool - AssertionError: Dtypes torch.int64 and torch.bool are not equal!
FAILED test/test_ops.py::TestCommonCPU::test_python_ref_torch_fallback__refs_diag_embed_cpu_bool - AssertionError: Dtypes torch.int64 and torch.bool are not equal!
FAILED test/test_ops.py::TestCommonCUDA::test_python_ref__refs_diag_embed_cuda_bool - AssertionError: Dtypes torch.int64 and torch.bool are not equal!
FAILED test/test_ops.py::TestCommonCUDA::test_python_ref_executor__refs_diag_embed_executor_aten_cuda_bool - AssertionError: Dtypes torch.int64 and torch.bool are not equal!
FAILED test/test_ops.py::TestCommonCUDA::test_python_ref_torch_fallback__refs_diag_embed_cuda_bool - AssertionError: Dtypes torch.int64 and torch.bool are not equal!
```

Reply:
you probably want to do `False if cond.dtype == bool else 0` for clarity. Otherwise passing `False` and having this casted to `0` automagically is too cryptic.